### ISSUE TITLE
WEBGPU: unexpected right shift when use ALLOW_MEMORY_GROWTH and MAXIMUM_MEMORY

### ISSUE THREAD

**Issue by xdestiny110** [emscripten-core/emscripten#20538]
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.47 (431685f05c67f0424c11473cc16798b9587bb536)

The following command
```
em++ -sALLOW_MEMORY_GROWTH=1 -sUSE_WEBGPU=1 -sMAXIMUM_MEMORY=4gb test.cpp
```
generates `a.out.js` as normal. But if we look into the generated `a.out.js`，
```js
function _wgpuBufferGetMappedRange(bufferId, offset, size) {
 bufferId >>>= 0;
 offset >>>= 0;
 size >>>= 0;
 var bufferWrapper = WebGPU.mgrBuffer.objects[bufferId];
 assert(typeof bufferWrapper != "undefined");
 if (size === 0) warnOnce("getMappedRange size=0 no longer means WGPU_WHOLE_MAP_SIZE");
 if (size == -1) size = undefined;
....
```
it performs a right shift operation on the size parameter at the beginning, which results in a non-negative number. However, the subsequent code logic contains a comparison where `size == -1` is checked. This means that size can never become undefined, consequently causing the underlying WebGPU to not perform as expected.
When I remove `ALLOW_MEMORY_GROWTH` or `MAXIMUM_MEMORY`, the right shift operator disappears.
This is `test.cpp` content
```cpp
#include <webgpu/webgpu_cpp.h>

int main() {
    wgpu::Buffer buffer;
    buffer.GetMappedRange();
    return 0;
}
```

**Comment by sbc100** [comment]
Yup, it looks like the `if (size == -1) size = undefined;` is probably wrong since the type of size is `size_t` which is an unsigned type.

**Comment by kainino0x** [comment]
Hm, we didn't test this with the right build config. This almost certainly explains why for so long I thought we had all our sentinel checks correct (checking for 0xFFFFFFFF (didn't work in 64bit)), then it somehow turned out they were all wrong (needed to check for -1).

> When I remove `ALLOW_MEMORY_GROWTH` or `MAXIMUM_MEMORY`, the right shift operator disappears.

@sbc100 how should we deal with this so it works in both configs? Should we just cast it like `if ((size | 0) == -1)`?

**Comment by sbc100** [comment]
This could be tricky.. because we convert `size_t` and pointer values to to JS numbers when they arrive in JS.  We have `{{{ POINTER_MAX }}}`, but that is defined as `MEMORY64 ? 'Number.MAX_SAFE_INTEGER' : '0xFFFFFFFF';`.    I'm not sure our current setup allows for a pointer value to be larger than `MAX_SAFE_INTEGER` in wasm64. 

What is the sentinal value defined as in C?

**Comment by kainino0x** [comment]
> What is the sentinal value defined as in C?

[SIZE_MAX](https://github.com/emscripten-core/emscripten/blob/4ecfddd1e31f5f4af56868df5c281c3650c8d8d5/system/include/webgpu/webgpu.h#L77), for this one which is size_t. (There are some others where it's UINT32_MAX or UINT64_MAX.)

**Comment by kainino0x** [comment]
This won't be fixed in Emscripten's current WebGPU bindings, but I believe it is fixed in `emdawnwebgpu` (see #23432) in https://crbug.com/389977397
